### PR TITLE
feat: Add Run columns to GetProjectColumns

### PR DESCRIPTION
**Pull request by AmanuelAaron**
## Description
Add support for Run Table Columns in GetProjectColumns endpoint

## Test Plan

integration tests should pass

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

## Ticket
ET-176

### CODE REVIEW

File: master/internal/api_project.go [hunk 1]
```diff
@@ -318,6 +318,267 @@ func (a *apiServer) getProjectColumnsByID(
...
+		JSONPath   string
+		MetricType string
+		Count      int32
+	}{}
+
+	if len(runIDs) > 0 {
+		subQuery := db.BunSelectMetricGroupNames().ColumnExpr(
+			`summary_metrics->jsonb_object_keys(summary_metrics)->
+		jsonb_object_keys(summary_metrics->jsonb_object_keys(summary_metrics))->>'type'
+		AS metric_type`).
```
Comment (NicholasBlaskey) [inline master/internal/api_project.go:483]: can we add a performance test covering this?

Comment (AmanuelAaron) [reply master/internal/api_project.go:483]: added one test for each table type.

Comment (NicholasBlaskey) [reply master/internal/api_project.go:483]: Do you know if we can improve this at all? It looks like an average of 50 seconds.

```
Get Project Columns for Runs Table {'test_name': 'Get Project Columns for Runs Table', 'avg': 51845.51096350322, 'min': 13163.07077, 'med': 60000.411748, 'max': 60001.197938, 'p90': 60000.985919599996, 'p95': 60001.1124485, 'passes': 64, 'fails': 91}
```

**Review by EmilyBonar — DISMISSED** [review]
Make sure you add the following columns:

- experimentProgress
- experimentName (move from `getProjectColumnsById` after rebase)
- experimentId
- isExpMultitrial
- parentArchived

**Review by johnkim-det — COMMENTED** [review]
Looks like a couple columns with Location Type `RUN` are returned even when the `tableType` is `EXPERIMENT`, seems like those should not be included?

Also, I don't know if product requirements ever specified which columns need to be displayed in the Runs view (maybe Piyush can confirm?), but based on the columns expected by the current front-end code, it looks like these columns Emily previously requested are still not returned when `tableType` is `RUN`:
- experimentProgress
- experimentId
- experimentName
  - I do see `name` returned, but I don't think we actually name runs themselves so preferably this would be explicit and be called `experimentName` instead.
  - Looks like this is returned when `tableType` is `EXPERIMENT` right now?
- isExpMultiTrial
- parentArchived
  - We also don't seem to return `archived` in the current `getProjectColumns` for Experiment List. I think we had WEB-1528 to add it to the Experiment List once we decided what to show in the cells. So we could hold off until then and add it for both `tableType`s, or if it's not a big lift, I don't think it would hurt to add it in the response for both while you're making changes to this endpoint anyway.

I also don't seem to have any metrics columns being returned (with location type `TRAINING`, `VALIDATIONS`, `CUSTOMMETRIC`), but I think that may be just because my local data is lacking. But just calling that out to make sure your testing checks that these are returned.

File: master/internal/api_project.go [hunk 2]
```diff
@@ -330,6 +459,129 @@ func (a *apiServer) getProjectColumnsByID(
...
+		return nil, err
+	}
+
+	summaryMetrics := []struct {
+		MetricName string
+		JSONPath   string
+		MetricType string
+	}{}
+
+	if len(hyperparameters) > 0 {
```
[the remark this answers is not in the document]
Comment (AmanuelAaron) [reply master/internal/api_project.go:503]: Talked with Piyush about this a while ago. It is rare for customers to have hyperparameters more than 1 depth (i.e hp.learning_rate). This migration does up to depth 2 (i.e hp.param1.param2). For future use though any depth can be added.

File: master/static/migrations/20240508134054_add-run-hparam-table.tx.up.sql [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+CASE WHEN s.type='string' THEN s.value::text ELSE NULL END as text_val,
+CASE WHEN s.type='boolean' THEN s.value::text::boolean ELSE NULL END as bool_val
+FROM (SELECT r.id, h.key, h.value, jsonb_typeof(h.value) as type FROM runs as r, jsonb_each(r.hparams) as h WHERE r.hparams is not NULL AND r.hparams !='null') as s WHERE s.type!='object';
+
+INSERT INTO run_hparams(run_id, hparam, number_val, text_val, bool_val)
+SELECT n.id as run_id, n.key as hparam,
+CASE WHEN n.type='number' THEN n.value::text::float ELSE NULL END as number_val,
+CASE WHEN n.type='string' THEN n.value::text ELSE NULL END as text_val,
+CASE WHEN n.type='boolean' THEN n.value::text::boolean ELSE NULL END as bool_val
+FROM (SELECT s.id, CONCAT(s.key, '.', nh.key) as key, nh.value,  jsonb_typeof(nh.value) as type FROM (SELECT r.id, h.key, h.value, jsonb_typeof(h.value) as type FROM runs as r, jsonb_each(r.hparams) as h WHERE r.hparams is not NULL AND r.hparams !='null') as s, jsonb_each(s.value) nh WHERE s.type='object') as n;
```
[the remark this answers is not in the document]
Comment (AmanuelAaron) [reply master/static/migrations/20240508134054_add-run-hparam-table.tx.up.sql:21]: Talked with Piyush about this a while ago. It is rare for customers to have hyperparameters more than 1 depth (i.e hp.learning_rate). This migration does up to depth 2 (i.e hp.param1.param2). For future use though any depth can be added.

File: master/internal/db/postgres_trial.go [hunk 4]
```diff
@@ -60,6 +82,117 @@ func AddTrial(ctx context.Context, trial *model.Trial, taskID model.TaskID) erro
...
+			conv := float64(val)
+			hp.NumberVal = &conv
+		case string:
+			hp.TextVal = &val
+		case bool:
+			hp.BoolVal = &val
+		case map[string]any:
+			nestedHParams, err := AddRunHParams(runID, v.(map[string]any), hpName+".")
+			if err != nil {
+				return hparamsModel, fmt.Errorf("failed to get nested hyperperameters for %s", hpName)
```
[the remark this answers is not in the document]
Comment (AmanuelAaron) [reply master/internal/db/postgres_trial.go]: Combined them into 1 function doing 1 loop instead of two.

File: master/static/migrations/20240508134054_add-run-hparam-table.tx.up.sql (hunk 3 above)
[the remark this answers is not in the document]
Comment (NicholasBlaskey) [reply master/static/migrations/20240508134054_add-run-hparam-table.tx.up.sql:21]: I would really like to avoid not supporting certain uses. If a customer relies on this feature it would be a bad experience. If it is decided that we don't support a use case then I think we would need to announce that in release notes and let customer success know.

Are we sure there is no way to support any nesting? Can we base this off what @gt2345 did in the past?
https://github.com/determined-ai/determined/blob/31bc08a5df3a1a42974c455039c48a783501de04/master/static/migrations/20230329193551_add-hyperparameters-metrics-name-view.tx.up.sql#L3

[the remark this answers is not in the document]
Comment (AmanuelAaron) [reply master/static/migrations/20240508134054_add-run-hparam-table.tx.up.sql]: Added a recursive query.

File: master/static/migrations/20240520104140_add-run-hparam-table.tx.up.sql [hunk 5]
```diff
@@ -0,0 +1,42 @@
+CREATE TABLE run_hparams (
```
Comment (NicholasBlaskey) [inline master/static/migrations/20240520104140_add-run-hparam-table.tx.up.sql:1]: we drop ```projects.hyperparameters``` and delete the associated code now that it is replaced by this?

Can do this in a follow up ticket if you want

File: master/internal/api_project.go [hunk 6]
```diff
@@ -28,6 +28,206 @@ import (
...
+	jsonb_object_keys(summary_metrics->jsonb_object_keys(summary_metrics))->>'type'
+	AS metric_type`).
+		Where(whereClause, bun.In(group)).Distinct()
+	trialsQuery := db.Bun().NewSelect().TableExpr("(?) AS stats", subQuery).
+		ColumnExpr("*").ColumnExpr(
+		"ROW_NUMBER() OVER(PARTITION BY json_path, metric_name order by metric_type) AS count").
+		Order("json_path").Order("metric_name")
+	err := trialsQuery.Scan(ctx, &summaryMetrics)
+	if err != nil {
+		return nil, err
```
Comment (NicholasBlaskey) [inline master/internal/api_project.go:186]: wrap this error and any error that comes from bun

File: master/internal/api_project.go [hunk 7]
```diff
@@ -330,6 +463,67 @@ func (a *apiServer) getProjectColumnsByID(
...
+
+	columns := defaultRunsTableColumns
+
+	hyperparameters := []struct {
+		WorkspaceID int
+		Hparam      string
+		Type        string
+	}{}
+
+	// get all runs in project
```
Comment (NicholasBlaskey) [inline master/internal/api_project.go:482]: nit: comments should be sentences starting with a capital letter and ending with a period.

File: master/internal/api_project.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (NicholasBlaskey) [reply master/internal/api_project.go:507]: technically you could have a project without hyperparameters but have them have summary metrics

I think its a premature optimization

File: performance/src/api_performance_tests.ts [hunk 8]
```diff
@@ -173,6 +173,20 @@ const getloadTests = (
...
     ),
+    test(
+      "Get Project Columns for Runs Table",
+      getRequest(
+        `/api/v1/projects/${sD?.workspace.projectId}/columns?table_type=TABLE_TYPE_RUN`,
+      ),
+      !!sD?.workspace.projectId,
+    ),
+    test(
+      "Get Project Columns for Experiments Table",
```
Comment (NicholasBlaskey) [inline performance/src/api_performance_tests.ts:184]: is this the same endpoint as https://github.com/determined-ai/determined/blob/805f2f4ea5af73b2f141f33a7771a3c5e255e8d9/performance/src/api_performance_tests.ts#L196C44-L196C53?

is it going to slow down all the other performance tests while we fix it or has it been improved?

**Review by NicholasBlaskey — APPROVED** [review]
lgtm

File: performance/src/api_performance_tests.ts (hunk 8 above)
Comment (AmanuelAaron) [reply performance/src/api_performance_tests.ts]: It is effectively the same call. I removed it as only the performance for the new columns have been improved.
